crypto/bn/asm/rsaz-avx512.pl

```diff
@@ -0,0 +1,684 @@
+# Almost Montgomery Multiplication for 20-digit number in radix 2^52
+#
+# void RSAZ_amm52x20(BN_ULONG *res,
+#                    const BN_ULONG *a,
+#                    const BN_ULONG *b,
+#                    const BN_ULONG *m,
+#                    BN_ULONG k0);
+###############################################################################
+{
+# input parameters
+my ($res,$a,$b,$m,$k0) = @_6_args_universal_ABI;
+
+my ($A0,$A1,$A2, $M0,$M1,$M2, $Bi,$Yi, $K0,$Rlo52) = map("%zmm$_",(16..25));
+my ($R0,$R1,$R2,$ZERO) = map("%zmm$_",(0..4));
+my $R2xmm = $R2 =~ s/%z/%x/r;
+
+$code.=<<___;
+.text
+
+.globl  RSAZ_amm52x20
+.type   RSAZ_amm52x20,\@function,5
+.align 32
+RSAZ_amm52x20:
+.cfi_startproc
+    endbranch
+    movl    \$15, %eax
+    kmovd   %eax, %k2                   # mask of 0-3 elements
+
+    vmovdqu64   ($a),    $A0            # load a
+    vmovdqu64   64($a),  $A1
+    vmovdqu64   128($a), ${A2}{%k2}{z}
+
+    vmovdqu64   ($m),    $M0            # load m
+    vmovdqu64   64($m),  $M1
+    vmovdqu64   128($m), ${M2}{%k2}{z}
+
+    movl    \$1, %eax
+    kmovd   %eax, %k1                   # mask of 0-th element
+    vpbroadcastq    $k0, $K0            # broadcast k0
+
+    leaq    160($b), %rax               # end of b (20-th digit)
+
+    vpxor   $R2xmm, $R2xmm, $R2xmm      # zeroing R0, R1, R2 accumulators
+    vmovdqa64   $R2, $R1
+    vmovdqa64   $R2, $R0
+    vmovdqa64   $R2, $ZERO
+
+.align 32
+.Lloop20:
+    vpbroadcastq ($b), $Bi              # Bi: broadcast b[i]
+    vmovdqa64   $ZERO, $Yi              # Yi = 0
+
+    vpmadd52luq $Bi, $A0, $R0           # R += A * Bi(lo)
+    vpmadd52luq $Bi, $A1, $R1
+    vpmadd52luq $Bi, $A2, $R2
+    addq    \$8, $b
+
+    vpmadd52luq $R0, $K0, $Yi           # Yi = R0 * k
+    vpermq  $Yi, $ZERO, $Yi             # broadcast Yi
+
+    vpmadd52luq $Yi, $M0, $R0           # R += M * Yi
+    vpmadd52luq $Yi, $M1, $R1
+    vpmadd52luq $Yi, $M2, $R2
+
+    vpsrlq  \$52, $R0, ${Rlo52}{%k1}{z} # Shift R
+    valignq \$1, $R0, $R1, $R0
+    valignq \$1, $R1, $R2, $R1
+    valignq \$1, $R2, $ZERO, $R2
+    vpaddq  $R0, $Rlo52, $R0
+
+    vpmadd52huq $Bi, $A0, $R0           # R += A * Bi(hi)
+    vpmadd52huq $Bi, $A1, $R1
+    vpmadd52huq $Bi, $A2, $R2
+
+    vpmadd52huq $Yi, $M0, $R0           # R += M * Yi
+    vpmadd52huq $Yi, $M1, $R1
+    vpmadd52huq $Yi, $M2, $R2
+
+    cmpq    %rax, $b                    # if not last digit
+    jne .Lloop20
+
+    vmovdqu64   $R0, ($res)             # Store de-normalized results
+    vmovdqu64   $R1, 64($res)
+    vmovdqu64   $R2, 128($res){%k2}
+
+    leaq    160($res), %r10             # end of res (20-th digit)
+
+    xorl    %eax, %eax
+    mov     \$0xfffffffffffff, %r11     # 52-bit mask
+
+.align 32
+.Lnorm:                                 # Normalize result
+    addq    ($res), %rax
+    movq    %rax, %rdx
+    andq    %r11, %rdx
+    movq    %rdx, ($res)
+    addq    \$8, $res
+    shrq    \$52, %rax
+    cmpq    $res, %r10
+    jne .Lnorm
```

Comment:
Currently not understanding the code, this looks like branches we want to avoid.

Reply:
On second look, this is probably fine.

Reply:
This is a normalization loop after multiplication. Note that 2^52 radix digits representation is used here, so upper 12 bits hold the "carry" that need to be reallocated to preserve original radix.